libraries/libstratosphere/source/settings/impl/settings_key_value_store.cpp

```diff
@@ -0,0 +1,1629 @@
+        /* Acquire exclusive access to global state. */
+        std::scoped_lock lk(g_key_value_store_mutex);
+
+        /* Get the key value store map. */
+        Map *map = nullptr;
+        R_TRY(GetKeyValueStoreMapForciblyForDebug(std::addressof(map)));
+        AMS_ASSERT(map != nullptr);
+
+        /* Iterate through each item. */
+        for (size_t i = 0; i < items_count; i++) {
+            const KeyValueStoreItemForDebug &item = items[i];
+
+            /* Get the map value for the item. */
+            MapValue map_value;
+            R_TRY(GetMapValueOfKeyValueStoreItemForDebug(std::addressof(map_value), item));
+
+            /* Ensure we free any buffers we allocate, if we fail. */
+            ON_SCOPE_EXIT { FreeMapValueToHeap(map_value); };
+
+            /* Ensure there is sufficient memory for two keys. */
+            R_UNLESS(GetHeapAllocatableSize() >= MapKey::MaxKeySize * 2, ResultSettingsItemKeyAllocationFailed());
+
+            /* Create the map key. */
+            MapKey map_key(item.key);
+
+            /* Replace the existing value in the map if it already exists. */
+            if (auto it = map->find(map_key); it != map->end()) {
+                /* Get the existing map value. */
+                MapValue *current_map_value = std::addressof(it->second);
+
+                /* Free the existing map value. */
+                FreeMapValueToHeap(*current_map_value);
+
+                /* Replace the existing map value. */
+                *current_map_value = map_value;
+            } else {
+                /* Ensure there is sufficient memory for the value. */
+                R_UNLESS(GetHeapAllocatableSize() >= MapValue::MaxValueSize, ResultSettingsItemValueAllocationFailed());
+            
+                /* Assign the map value to the map key in the map. */
+                (*map)[std::move(map_key)] = map_value;
+            }
+
+            /* Ensure we don't free the value buffers we added. */
+            map_value.current_value = nullptr;
+            map_value.default_value = nullptr;
+        }
+
+        return ResultSuccess();
+    }
+
+    Result AdvanceKeyValueStoreKeyIterator(KeyValueStoreKeyIterator *out) {
+        /* Check preconditions. */
+        AMS_ASSERT(out != nullptr);
+        AMS_ASSERT(out->name_size > 0);
+        AMS_ASSERT(out->name_size < out->key_size);
+        AMS_ASSERT(out->map_key != nullptr);
+
+        /* Acquire exclusive access to global state. */
+        std::scoped_lock lk(g_key_value_store_mutex);
+
+        /* Get the key value store map. */
+        Map *map = nullptr;
+        R_TRY(GetKeyValueStoreMap(std::addressof(map)));
+        AMS_ASSERT(map != nullptr);
+
+        /* Ensure there is sufficient memory for two keys. */
+        R_UNLESS(GetHeapAllocatableSize() >= MapKey::MaxKeySize * 2, ResultSettingsItemKeyAllocationFailed());
+
+        /* Locate the iterator's current key. */
+        auto it = map->find(MapKey(out->map_key, out->key_size - 1)); 
```

Comment:
`Map::const_iterator it = map->find(MapKey(out->map_key, static_cast<s32>(out->key_size) - 1))`